src/core/morphing.js

```diff
@@ -0,0 +1,110 @@
+import { Idiomorph } from "idiomorph/dist/idiomorph.esm.js"
+import { FrameElement } from "../elements/frame_element"
+import { dispatch } from "../util"
+
+export function morphRefresh(currentElement, newElement) {
+  idiomorph(currentElement, newElement, {
+    shouldSkipMorphing: element => canRefreshFrame(element)
+  })
+
+  for (const frame of document.querySelectorAll("turbo-frame")) {
+    if (canRefreshFrame(frame)) refreshFrame(frame)
+  }
+
+  dispatch("turbo:morph", { detail: { currentElement, newElement } })
+}
+
+export function morphFrames(currentElement, newElement) {
+  dispatch("turbo:before-frame-morph", {
+    target: currentElement,
+    detail: { currentElement, newElement }
+  })
+
+  morphChildren(currentElement, newElement)
+}
+
+export function morphChildren(currentElement, newElement) {
+  idiomorph(currentElement, newElement.children, {
+    options: {
+      morphStyle: "innerHTML"
+    }
+  })
+}
+
+export function morphElements(currentElement, newElement) {
+  idiomorph(currentElement, newElement)
+}
+
+const defaultOptions = {
+  morphStyle: "outerHTML"
+}
+
+function idiomorph(currentElement, newElement, delegate = {}) {
+  const options = delegate.options || {}
+  const callbacks = new IdiomorphDelegate(delegate)
+
+  Idiomorph.morph(currentElement, newElement, {
+    ...defaultOptions,
+    ...options,
+    callbacks
+  })
+}
+
+class IdiomorphDelegate {
```

Comment:
Suggestion: rename this to `IdiomorphCallbacks`. I know we use `delegate` a lot in Turbo's codebase, but I must say it's my least favorite part of the code. I think *callbacks* would be a better pick to clarify what the object does. I found *delegate* a bit void of meaning in general.